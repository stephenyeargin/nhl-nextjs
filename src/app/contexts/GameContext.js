import React, { createContext, useState, useEffect, useContext } from 'react';
import { PropTypes } from 'prop-types';
import { GAME_STATES } from '../utils/constants';

const GameContext = createContext();

export const useGameContext = () => useContext(GameContext);

export const GameProvider = ({ gameId, children }) => {
  const [gameData, setGameData] = useState(null);
  const [gameState, setGameState] = useState(null);
  const [pageError, setPageError] = useState(null);

  useEffect(() => {
    let intervalId;

    const fetchGameData = async () => {
      let game, rightRail, story;
      try {
        const gameResponse = await fetch(`/api/nhl/gamecenter/${gameId}/landing`, { cache: 'no-store' });
        const rightRailResponse = await fetch(`/api/nhl/gamecenter/${gameId}/right-rail`, { cache: 'no-store' });
        const storyResponse = await fetch(`/api/nhl/wsc/game-story/${gameId}`, { cache: 'no-store' });

        if (!gameResponse.ok || !rightRailResponse.ok || !storyResponse.ok) {
          throw new Error('Failed to fetch game data');
        }

        game = await gameResponse.json();
        rightRail = await rightRailResponse.json();
        story = await storyResponse.json();
      } catch (error) {
        setPageError({ message: 'Failed to load the game data. Please try again later.', error });
        console.error('Error fetching game data:', error);
      }

      const { homeTeam, awayTeam, gameDate, venue, venueLocation, summary, matchup } = game || {};
      setGameData({ homeTeam, awayTeam, gameDate, venue, venueLocation, summary, matchup, game, rightRail, story });
      setGameState(game ? game.gameState : null);
    };

    // Initial fetch
    fetchGameData();
   
    // Polling interval (only if the game is in progress)
    if (['PRE', 'LIVE', 'CRIT'].includes(gameState)) {
      intervalId = setInterval(() => {
        fetchGameData();
      }, 20000); // 20 seconds
    }
    
    return () => clearInterval(intervalId);
  }, [gameId, gameState]); // only re-run if gameState or gameId changes

  // Set page title
  window.document.title = gameData ? `${gameData.homeTeam.abbrev} vs ${gameData.awayTeam.abbrev}` : 'NHL Game';
  if (gameData && gameState !== 'FUT') {
    const { homeTeam, awayTeam } = gameData;
    window.document.title = `${awayTeam.abbrev} (${awayTeam.score}) vs ${homeTeam.abbrev} (${homeTeam.score}) - ${GAME_STATES[gameState]}`;
  }

  return (
    <GameContext.Provider value={{ gameData, gameState, pageError }}>
      {children}
    </GameContext.Provider>
  );
};

GameProvider.propTypes = {
  gameId: PropTypes.string.isRequired,
  children: PropTypes.node.isRequired,
};

export default GameProvider;
