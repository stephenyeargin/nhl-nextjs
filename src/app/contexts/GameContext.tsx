import React, { createContext, useState, useEffect, useContext } from 'react';
import { GAME_STATES } from '@/app/utils/constants';
import { formatHeadTitle } from '@/app/utils/formatters';

// --- Domain shapes (partial) ---
interface TeamSide { abbrev: string; score?: number; [k: string]: any }
interface GameLanding {
  gameState?: string;
  gameScheduleState?: string;
  homeTeam: TeamSide;
  awayTeam: TeamSide;
  gameDate?: string;
  venue?: any;
  venueLocation?: any;
  summary?: any;
  matchup?: any;
  [k: string]: any;
}
interface RightRailData { [k: string]: any }
interface GameStoryData { [k: string]: any }

interface GameDataShape {
  homeTeam?: TeamSide;
  awayTeam?: TeamSide;
  gameDate?: string;
  venue?: any;
  venueLocation?: any;
  summary?: any;
  matchup?: any;
  game?: GameLanding | null;
  rightRail?: RightRailData | null;
  story?: GameStoryData | null;
}

interface PageError { message: string; error: { status?: number } | any }

interface GameContextValue {
  gameData: GameDataShape | null;
  gameState: string | null;
  pageError: PageError | null;
}

export const GameContext = createContext<GameContextValue | undefined>(undefined);
export const useGameContext = (): GameContextValue => {
  const ctx = useContext(GameContext);
  if (!ctx) {
    throw new Error('useGameContext must be used within a GameProvider');
  }

  return ctx;
};

const GAME_REFRESH_TTL = 15 * 1000; // 15 seconds

interface GameProviderProps { gameId: string; children: React.ReactNode }

export const GameProvider: React.FC<GameProviderProps> = ({ gameId, children }) => {
  const [gameData, setGameData] = useState<GameDataShape | null>(null);
  const [gameState, setGameState] = useState<string | null>(null);
  const [gameScheduleState, setGameScheduleState] = useState<string | null>(null);
  const [pageError, setPageError] = useState<PageError | null>(null);

  useEffect(() => {
    let intervalId: ReturnType<typeof setInterval> | undefined;

    const fetchGameData = async () => {
      let game: GameLanding | undefined; let rightRail: RightRailData | undefined; let story: GameStoryData | undefined;
      try {
        const [ gameResponse, rightRailResponse, storyResponse ] = await Promise.all([
          fetch(`/api/nhl/gamecenter/${gameId}/landing`, { cache: 'no-store' }),
          fetch(`/api/nhl/gamecenter/${gameId}/right-rail`, { cache: 'no-store' }),
          fetch(`/api/nhl/wsc/game-story/${gameId}`, { cache: 'no-store' })
        ]);

        if (!gameResponse.ok || !rightRailResponse.ok || !storyResponse.ok) {
          const notFound = gameResponse.status === 404;
          setPageError({ message: notFound ? 'Game not found.' : 'Failed to load the game data. Please try again later.', error: { status: notFound ? 404 : 500 } });
          throw new Error('Error loading game data.');
        }

        game = await gameResponse.json();
        rightRail = await rightRailResponse.json();
        story = await storyResponse.json();

        // Set page title
        if (game?.homeTeam?.abbrev && game?.awayTeam?.abbrev) {
          formatHeadTitle(`${game.homeTeam.abbrev} vs. ${game.awayTeam.abbrev}`);
          if (game.gameState && !['FUT', 'PRE'].includes(game.gameState)) {
            const readable = GAME_STATES[game.gameState as keyof typeof GAME_STATES] || game.gameState;
            formatHeadTitle(`${game.awayTeam.abbrev} (${game.awayTeam.score}) vs. ${game.homeTeam.abbrev} (${game.homeTeam.score}) - ${readable}`);
          }
        }
      } catch (error) {
        // Already captured high-level error above; just log for diagnostics
        // eslint-disable-next-line no-console
        console.error('Error fetching game data:', error);
      }

      const { homeTeam, awayTeam, gameDate, venue, venueLocation, summary, matchup } = game || {} as GameLanding;
      setGameData({ homeTeam, awayTeam, gameDate, venue, venueLocation, summary, matchup, game: game || null, rightRail: rightRail || null, story: story || null });
      setGameState(game ? game.gameState || null : null);
      setGameScheduleState(game ? game.gameScheduleState || null : null);
    };

    fetchGameData();

    if (!['OFF'].includes(gameState || '') && gameScheduleState === 'OK') {
      intervalId = setInterval(fetchGameData, GAME_REFRESH_TTL);
    }

    return () => {
      if (intervalId) {
        clearInterval(intervalId);
      }
    };
  }, [gameId, gameState, gameScheduleState]);

  const value: GameContextValue = { gameData, gameState, pageError };

  return (
    <GameContext.Provider value={value}>
      {children}
    </GameContext.Provider>
  );
};

export default GameProvider;
