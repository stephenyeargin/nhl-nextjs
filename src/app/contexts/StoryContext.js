import React, { createContext, useState, useEffect, useContext } from 'react';
import PropTypes from 'prop-types';
import { formatHeadTitle, formatLocalizedDate } from '../utils/formatters';

// Context & Custom Hook
const StoryContext = createContext();
export const useStoryContext = () => useContext(StoryContext);

// Helper function for fetching data
const fetchJsonData = async (url) => {
  const response = await fetch(url, { cache: 'no-store' });
  if (!response.ok) {
    throw new Error(`Failed to fetch data from ${url}`);
  }
  
  return response.json();
};

// Fetch story data based on the storyId
const fetchStory = async (storyId) => {
  const isLegacyStory = /^c-\d+$/.test(storyId);
  if (isLegacyStory) {
    // Handle legacy story by looking up homebase content
    const homebaseContent = await fetchJsonData(
      `https://forge-dapi.d3.nhle.com/v2/content/en-us/stories/?fields.homebaseId=${storyId.replace('c-', '')}`
    );
    
    return homebaseContent.items[0].slug;
  }

  // Otherwise, fetch the full story data
  return await fetchJsonData(`https://forge-dapi.d3.nhle.com/v2/content/en-us/stories/${storyId}`);
};

// Fetch sidebar stories
const fetchSidebarStories = async () => {
  return await fetchJsonData(
    'https://forge-dapi.d3.nhle.com/v2/content/en-us/stories?tags.slug=news&context.slug=nhl&$limit=5'
  );
};

// Fetch game data based on the story's game tag
const fetchGameData = async (gameId) => {
  return await fetchJsonData(`/api/nhl/gamecenter/${gameId}/landing`);
};

// Story Provider Component
export const StoryProvider = ({ storyId, children }) => {
  const [story, setStory] = useState({});
  const [game, setGame] = useState({});
  const [sidebarStories, setSidebarStories] = useState([]);
  const [pageError, setPageError] = useState(null);

  useEffect(() => {
    const loadData = async () => {
      try {
        // Step 1: Fetch story
        const storyResponse = await fetchStory(storyId);

        // Step 2: Handle redirect for legacy story if needed
        if (typeof storyResponse === 'string') {
          return window.location.replace(`/news/${storyResponse}`);
        }

        // Step 3: Fetch sidebar stories
        const topStories = await fetchSidebarStories();

        // Step 4: Fetch game data if story has a game tag
        const gameTag = storyResponse.tags.find((t) => t.externalSourceName === 'game');
        if (gameTag) {
          const gameId = gameTag.extraData.gameId;
          const gameResponse = await fetchGameData(gameId);
          setGame(gameResponse);
        }

        // Step 5: Update state with fetched data
        setSidebarStories(topStories);
        setStory(storyResponse);

        // Step 6: Set the page title
        formatHeadTitle(`${formatLocalizedDate(story.contentDate)}: ${story.headline || story.title}`);

      } catch (error) {
        setPageError({ message: 'Failed to load story. Please try again later.', error });
        console.error('Error fetching story data:', error);
      }
    };

    loadData();
  }, [storyId, story]); // only re-run if story/storyId changes

  return (
    <StoryContext.Provider value={{ story, game, sidebarStories, pageError }}>
      {children}
    </StoryContext.Provider>
  );
};

// Prop Types Validation
StoryProvider.propTypes = {
  storyId: PropTypes.string.isRequired,
  children: PropTypes.node.isRequired,
};

export default StoryProvider;
