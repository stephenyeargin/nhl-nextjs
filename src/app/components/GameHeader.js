'use client';

import React, { useState, useEffect, useRef } from 'react';
import Link from 'next/link';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faPauseCircle, faWarning, faBan, faClock } from '@fortawesome/free-solid-svg-icons';
import GameClock from './GameClock';
import { formatLocalizedDate, formatLocalizedTime, formatPeriodLabel } from '../utils/formatters';
import TeamLogo from './TeamLogo';
import { PropTypes } from 'prop-types';
import SirenOnSVG from '@/app/assets/siren-on-solid.svg';
import Image from 'next/image';
import { useGameContext } from '../contexts/GameContext';
import { getTeamDataByAbbreviation } from '../utils/teamData';
import GameHeaderSkeleton from './GameHeaderSkeleton';

const GameHeader = () => {
  const [isSticky, setIsSticky] = useState(false);
  const stickyRef = useRef(null);

  const { gameData } = useGameContext();

  useEffect(() => {
    const handleScroll = () => {
      if (stickyRef.current) {
        const { top } = stickyRef.current.getBoundingClientRect();
        setIsSticky(top <= 0); // If the element's top is less than or equal to 0, it's sticky
      }
    };
    window.addEventListener('scroll', handleScroll);

    return () => {
      window.removeEventListener('scroll', handleScroll);
    };
  }, []);

  if (!gameData || !gameData.game) {
    return <GameHeaderSkeleton />;
  }

  // Destructure data for rendering
  const { game } = gameData;
  const { venue, venueLocation, awayTeam, homeTeam, gameState, gameScheduleState, ifNecessary, periodDescriptor, situation, clock, startTimeUTC } = game;

  awayTeam.data = getTeamDataByAbbreviation(awayTeam.abbrev, false);
  homeTeam.data = getTeamDataByAbbreviation(homeTeam.abbrev, true);

  let gameHeaderClass = 'grid grid-cols-12 my-5 border rounded-lg shadow-sm py-4 items-center';
  if (isSticky) {
    gameHeaderClass = 'grid grid-cols-12 items-center md:sticky top-0 pt-5 z-10 border rounded-b-lg shadow-md pb-5';
  }
  if (gameState === 'CRIT') {
    gameHeaderClass += ' border-red-900';
  }

  const gameHeaderStyle = { background: 'var(--background)' };
  gameHeaderStyle.borderLeft = `solid 10px ${awayTeam.data.teamColor}`;
  gameHeaderStyle.borderRight = `solid 10px ${homeTeam.data.teamColor}`;

  const teamHasRecentGoal = (teamAbbrev, game) => {
    // No goals if game isn't live
    if (!['LIVE', 'CRIT'].includes(game.gameState)) {
      return false;
    }

    // No goals during intermission
    if (game.clock.inIntermission) {
      return false;
    }

    // Retrieve the current period number
    const currentPeriod = game.periodDescriptor.number;

    // Find the most recent goal scored by the specified team in the current period
    const currentPeriodGoals = game.summary.scoring
      .filter(period => period.periodDescriptor.number === currentPeriod)
      .flatMap(period => period.goals)
      .filter(goal => goal.teamAbbrev.default === teamAbbrev);

    if (currentPeriodGoals.length === 0) {
      return false; // No goals scored by the specified team in the current period
    }

    // Get the most recent goal's time in the period
    const mostRecentGoal = currentPeriodGoals[currentPeriodGoals.length - 1];
    const goalTimeStr = mostRecentGoal.timeInPeriod;
    const [goalMinutes, goalSeconds] = goalTimeStr.split(':').map(Number);
    const goalTime = goalMinutes * 60 + goalSeconds; // Convert to total seconds

    // Get the remaining time in the period
    const remainingTimeStr = game.clock.timeRemaining;
    const [remainingMinutes, remainingSeconds] = remainingTimeStr.split(':').map(Number);
    const remainingTime = remainingMinutes * 60 + remainingSeconds; // Convert to total seconds

    // Calculate the elapsed time in the period
    const totalPeriodTime = 20 * 60; // Assume a 20-minute period in seconds
    const elapsedTime = totalPeriodTime - remainingTime;

    // Check if the goal time is within 5 seconds of the elapsed time
    const timeDifference = Math.abs(elapsedTime - goalTime);

    return timeDifference <= 5;
  };

  return (
    <>
      {game.specialEvent && (
        <>
          {game.specialEvent.lightLogoUrl?.default ? (
            <div className="flex justify-center py-5">
              <Image
                src={game.specialEvent.lightLogoUrl.default}
                width={300}
                height={100}
                alt="Event Logo"
                title={game.specialEvent.name.default}
                className="w-full md:w-1/3"
              />
            </div>
          ) : (
            <div className="my-4 text-center font-bold text-gray-500">&#9734;&#9734;&#9734; {game.specialEvent.name.default} &#9734;&#9734;&#9734;</div>
          )}
        </>
      )}
      <div className={gameHeaderClass} ref={stickyRef} style={gameHeaderStyle}>
        <div className="col-span-3 flex flex-wrap mx-auto gap-2 items-center justify-center">
          <div>
            <TeamLogo
              team={awayTeam.abbrev}
              src={awayTeam.logo}
              alt={awayTeam.commonName.default}
              className="w-20 h-20 mx-auto"
            />
          </div>
          <div>
            <Link href={`/team/${awayTeam.abbrev}`}>
              <div className="text-xl font-black block md:hidden">{awayTeam.abbrev}</div>
              <div className="text-lg hidden md:block">
                <div className="text-sm">{awayTeam.placeName.default}</div>
                <div className="text-xl font-black">{awayTeam.commonName.default.replace(awayTeam.placeName.default, '')}</div>
              </div>
            </Link>
            {gameState !== 'FUT' ? (
              <div className="text-sm">SOG: {awayTeam.sog || 0}</div>
            ) : (
              <div className="text-sm">{awayTeam.record}</div>
            )}
          </div>
        </div>
        <div className={`col-span-2 flex flex-wrap justify-center items-center text-center text-5xl md:text-7xl font-black ${awayTeam.score < homeTeam.score && ['FINAL', 'OFF'].includes(gameState) ? 'opacity-50' : ''}`}>
          {situation?.awayTeam.situationDescriptions?.map((code) => (
            <span key={code} className="mx-1 text-lg rounded text-white bg-red-900 p-1 uppercase">{code}</span>
          ))}
          <span>{awayTeam.score}</span>
          {teamHasRecentGoal(awayTeam.abbrev, game) && (
            <Image src={SirenOnSVG} height="256" width="256" className="h-10 w-10 ml-2 animate-pulse" alt="Goal" />
          )}
        </div>
        <div className="col-span-2 text-center content-middle">
          {game.gameType === 1 && (
            <div className="text-xs my-1 p-1 border rounded font-bold uppercase">Preseason</div>
          )}
          {game.gameType === 3 && (
            <div className="text-xs my-1 p-1 border rounded font-bold uppercase">Stanley Cup Playoffs</div>
          )}
          <div className="text-xs my-1">{venue.default}, {venueLocation.default}</div>
          {(gameState === 'LIVE' || gameState === 'CRIT') && (
            <div className="my-3 flex flex-wrap justify-center">
              <span className="font-md font-medium px-2 py-1 bg-red-900 text-white rounded mr-2 uppercase text-nowrap">
                {formatPeriodLabel(periodDescriptor)}
                {clock.inIntermission ? ' INT' : ''}
              </span>
              {periodDescriptor.periodType !== 'SO' && (
                <span className="font-bold text-xl">
                  <GameClock timeRemaining={clock.timeRemaining} running={clock?.inIntermission || clock?.running} />
                  {!clock.inIntermission && !clock.running && (<FontAwesomeIcon icon={faPauseCircle} className="ml-1" />)}
                </span>
              )}
            </div>
          )}
          {(gameState === 'FINAL' || gameState === 'OFF') && (
            <div className="text-center my-2 uppercase">
              <span className="text-xs md:text-sm font-medium px-2 py-1 bg-slate-100 dark:text-black text-nowrap rounded">
              Final
                {periodDescriptor.periodType	!== 'REG' && `/${periodDescriptor.periodType}`}
              </span>
            </div>
          )}
          {['FUT', 'PRE'].includes(gameState) && (
            <div className="my-1">
              {gameScheduleState !== 'TBD' ? (
                <span className="text-xs md:text-sm font-medium px-2 py-1 bg-slate-100 text-black rounded uppercase text-nowrap">{formatLocalizedTime(startTimeUTC)}</span>
              ) : (
                <span className="text-xs md:text-sm font-medium px-2 py-1 bg-slate-100 text-black rounded uppercase text-nowrap">
                  {ifNecessary ? (
                    <span className="text-xs md:text-sm font-medium px-2 py-1 bg-slate-100 text-black rounded uppercase text-nowrap">(If Necessary)</span>
                  ) : (
                    <span className="text-xs md:text-sm font-medium px-2 py-1 bg-slate-100 text-black rounded uppercase text-nowrap">TBD</span>
                  )}
                </span>
              )}
            </div>
          )}
          {gameState === 'PRE' && (
            <div className="my-1 text-nowrap">
              <span className="text-sm font-medium px-2 py-1 bg-slate-900 text-white rounded uppercase"><FontAwesomeIcon icon={faClock} fixedWidth /> Pregame</span>
            </div>
          )}
          {gameScheduleState === 'CNCL' && (
            <div className="my-1 text-nowrap">
              <span className="text-sm font-medium px-2 py-1 bg-yellow-500 text-black rounded uppercase"><FontAwesomeIcon icon={faBan} fixedWidth /> Cancelled</span>
            </div>
          )}
          {gameScheduleState === 'PPD' && (
            <div className="my-1 text-nowrap">
              <span className="text-sm font-medium px-2 py-1 bg-yellow-500 text-black rounded uppercase"><FontAwesomeIcon icon={faWarning} fixedWidth /> Postponed</span>
            </div>
          )}
          {(situation && (situation.awayTeam.strength !== 5 || situation?.homeTeam.strength) !== 5) && (
            <div className="my-2">
              {situation?.timeRemaining && (
                <div>
                  <span className="text-sm font-medium px-2 py-1 border text-slate-900 dark:text-slate-100 rounded">
                    <GameClock timeRemaining={situation?.timeRemaining} running={clock?.running && !clock?.inIntermission} />
                  </span>
                </div>
              )}
              <span className="text-sm md:text-sm font-bold px-2 py-1 text-red-900 rounded uppercase text-nowrap">
                {situation?.awayTeam.strength}-on-{situation?.homeTeam.strength}
              </span>
            </div>
          )}
          {!['LIVE', 'CRIT'].includes(gameState) && (
            <div className="text-xs my-2">
              <div>{formatLocalizedDate(startTimeUTC, 'MMMM D, YYYY')}</div>
            </div>
          )}
        </div>
        <div className={`col-span-2 flex flex-wrap justify-center items-center text-center text-5xl md:text-7xl font-black ${awayTeam.score > homeTeam.score && ['FINAL', 'OFF'].includes(gameState) ? 'opacity-50' : ''}`}>
          {teamHasRecentGoal(homeTeam.abbrev, game) && (
            <Image src={SirenOnSVG} height="256" width="256" className="h-10 w-10 mr-2 animate-pulse" alt="Goal" />
          )}
          <span>{homeTeam.score}</span>
          {situation?.homeTeam.situationDescriptions?.map((code) => (
            <span key={code} className="mx-2 text-lg rounded text-white bg-red-900 p-1 uppercase">{code}</span>
          ))}
        </div>
        <div className="col-span-3 flex flex-wrap mx-auto gap-2 items-center justify-center">
          <div className="text-right order-2 sm:order-1">
            <Link href={`/team/${homeTeam.abbrev}`}>
              <div className="text-xl font-black block md:hidden">{homeTeam.abbrev}</div>
              <div className="text-lg hidden md:block">
                <div className="text-sm">{homeTeam.placeName.default}</div>
                <div className="text-xl font-black">{homeTeam.commonName.default.replace(homeTeam.placeName.default, '')}</div>
              </div>
            </Link>
            {gameState !== 'FUT' ? (
              <div className="text-sm">SOG: {homeTeam.sog || 0}</div>
            ) : (
              <div className="text-sm">{homeTeam.record}</div>
            )}
          </div>
          <div className="text-right order-1 sm:order-2">
            <TeamLogo
              team={homeTeam.abbrev}
              src={homeTeam.logo}
              alt={homeTeam.commonName.default}
              className="w-20 h-20 mx-auto"
            />
          </div>
        </div>
      </div>
    </>
  );
};

GameHeader.propTypes = {
  game: PropTypes.object.isRequired,
};

export default GameHeader;
