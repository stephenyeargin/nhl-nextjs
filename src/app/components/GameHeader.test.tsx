import React from 'react';
import { render, screen } from '@testing-library/react';
import GameHeader from './GameHeader';
import { GameContext } from '../contexts/GameContext';

jest.mock('./TeamLogo', () => {
  const Logo = ({ team, alt, className }: any) => (
    <span
      role="img"
      aria-label={alt || team}
      data-team={team}
      data-mock="team-logo"
      className={className}
    />
  );
  (Logo as any).displayName = 'TeamLogoMock';

  return Logo;
});

const teamBase = (abbrev: string, score: number) => ({
  abbrev,
  score,
  logo: 'logo.svg',
  sog: 10,
  record: '1-0-0',
  placeName: { default: `${abbrev} City` },
  commonName: { default: `${abbrev} City ${abbrev}` },
});

const baseGame = (overrides: any = {}) => ({
  gameState: 'OFF',
  gameScheduleState: 'OK',
  homeTeam: teamBase('HOM', 3),
  awayTeam: teamBase('AWY', 2),
  venue: { default: 'Arena' },
  venueLocation: { default: 'Somewhere' },
  periodDescriptor: { number: 3, periodType: 'REG' },
  situation: {
    awayTeam: { strength: 5, situationDescriptions: [] },
    homeTeam: { strength: 5, situationDescriptions: [] },
    timeRemaining: '00:00',
  },
  clock: { timeRemaining: '00:00', inIntermission: false, running: false },
  startTimeUTC: '2024-03-10T12:00:00Z',
  summary: { scoring: [] },
  ...overrides,
});

const ctx = (gameOverrides: any = {}) => {
  const game = baseGame(gameOverrides);

  return {
    gameData: {
      game,
      homeTeam: game.homeTeam,
      awayTeam: game.awayTeam,
    },
    gameState: game.gameState,
    pageError: null,
  } as any;
};

describe('GameHeader', () => {
  it('renders team names (baseline OFF state)', () => {
    const { asFragment } = render(
      <GameContext.Provider value={ctx()}>
        <GameHeader />
      </GameContext.Provider>
    );
    expect(screen.getAllByText(/AWY/)[0]).toBeInTheDocument();
    expect(screen.getAllByText(/HOM/)[0]).toBeInTheDocument();
    expect(screen.getByText('Final')).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it('shows preseason badge when gameType = 1', () => {
    render(
      <GameContext.Provider value={ctx({ gameType: 1, gameState: 'FUT' })}>
        <GameHeader />
      </GameContext.Provider>
    );
    expect(screen.getByText(/Preseason/i)).toBeInTheDocument();
  });

  it('shows playoffs badge when gameType = 3', () => {
    render(
      <GameContext.Provider value={ctx({ gameType: 3, gameState: 'FUT' })}>
        <GameHeader />
      </GameContext.Provider>
    );
    expect(screen.getByText(/Stanley Cup Playoffs/i)).toBeInTheDocument();
  });

  it('shows Pregame state badge for PRE gameState', () => {
    render(
      <GameContext.Provider value={ctx({ gameState: 'PRE' })}>
        <GameHeader />
      </GameContext.Provider>
    );
    expect(screen.getByText(/Pregame/i)).toBeInTheDocument();
  });

  it('shows TBD when future game schedule is TBD', () => {
    render(
      <GameContext.Provider
        value={ctx({
          gameState: 'FUT',
          gameScheduleState: 'TBD',
          startTimeUTC: '2024-03-10T12:00:00Z',
        })}
      >
        <GameHeader />
      </GameContext.Provider>
    );
    expect(screen.getAllByText(/TBD/i)[0]).toBeInTheDocument();
  });

  it('shows situation strength when not 5-on-5', () => {
    render(
      <GameContext.Provider
        value={ctx({
          gameState: 'LIVE',
          periodDescriptor: { number: 1, periodType: 'REG' },
          situation: {
            awayTeam: { strength: 3, situationDescriptions: ['PP'] },
            homeTeam: { strength: 5, situationDescriptions: [] },
            timeRemaining: '01:30',
          },
          clock: { timeRemaining: '10:00', inIntermission: false, running: false },
        })}
      >
        <GameHeader />
      </GameContext.Provider>
    );
    expect(screen.getByText(/3-on-5/)).toBeInTheDocument();
  });

  it('shows goal siren when recent goal scored', () => {
    // Elapsed 5 seconds into period -> goal at 00:05 counts as recent
    render(
      <GameContext.Provider
        value={ctx({
          gameState: 'LIVE',
          periodDescriptor: { number: 1, periodType: 'REG' },
          clock: { timeRemaining: '19:55', inIntermission: false, running: false },
          summary: {
            scoring: [
              {
                periodDescriptor: { number: 1 },
                goals: [{ teamAbbrev: { default: 'AWY' }, timeInPeriod: '00:05' }],
              },
            ],
          },
        })}
      >
        <GameHeader />
      </GameContext.Provider>
    );
    expect(screen.getAllByAltText('Goal')[0]).toBeInTheDocument();
  });

  it('shows cancelled badge for CNCL schedule state', () => {
    render(
      <GameContext.Provider value={ctx({ gameScheduleState: 'CNCL' })}>
        <GameHeader />
      </GameContext.Provider>
    );
    expect(screen.getByText(/Cancelled/i)).toBeInTheDocument();
  });

  it('shows postponed badge for PPD schedule state', () => {
    render(
      <GameContext.Provider value={ctx({ gameScheduleState: 'PPD' })}>
        <GameHeader />
      </GameContext.Provider>
    );
    expect(screen.getByText(/Postponed/i)).toBeInTheDocument();
  });
});
