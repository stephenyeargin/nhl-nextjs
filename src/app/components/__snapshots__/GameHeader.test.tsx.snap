// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameHeader renders team names (baseline OFF state) 1`] = `
<DocumentFragment>
  <div
    class="grid grid-cols-12 my-5 border rounded-lg shadow-sm py-4 items-center"
    style="border-left: 10px solid #FFB81C; border-right: 10px solid #3A5DAE;"
  >
    <div
      class="col-span-3 flex flex-wrap mx-auto gap-2 items-center justify-center"
    >
      <div>
        <span
          aria-label="AWY City AWY"
          class="w-20 h-20 mx-auto"
          data-mock="team-logo"
          data-team="AWY"
          role="img"
        />
      </div>
      <div>
        <a
          href="/team/AWY"
        >
          <div
            class="text-xl font-black block md:hidden"
          >
            AWY
          </div>
          <div
            class="text-lg hidden md:block"
          >
            <div
              class="text-sm"
            >
              AWY City
            </div>
            <div
              class="text-xl font-black"
            >
               AWY
            </div>
          </div>
        </a>
        <div
          class="text-sm"
        >
          SOG: 10
        </div>
      </div>
    </div>
    <div
      class="col-span-2 flex flex-wrap justify-center items-center text-center text-5xl md:text-7xl font-black opacity-50"
    >
      <span>
        2
      </span>
    </div>
    <div
      class="col-span-2 text-center content-middle"
    >
      <div
        class="text-xs my-1"
      >
        Arena, Somewhere
      </div>
      <div
        class="text-center my-2 uppercase"
      >
        <span
          class="text-xs md:text-sm font-medium px-2 py-1 bg-slate-100 dark:text-black text-nowrap rounded"
        >
          Final
        </span>
      </div>
      <div
        class="text-xs my-2"
      >
        <div>
          March 10, 2024
        </div>
      </div>
    </div>
    <div
      class="col-span-2 flex flex-wrap justify-center items-center text-center text-5xl md:text-7xl font-black "
    >
      <span>
        3
      </span>
    </div>
    <div
      class="col-span-3 flex flex-wrap mx-auto gap-2 items-center justify-center"
    >
      <div
        class="text-right order-2 sm:order-1"
      >
        <a
          href="/team/HOM"
        >
          <div
            class="text-xl font-black block md:hidden"
          >
            HOM
          </div>
          <div
            class="text-lg hidden md:block"
          >
            <div
              class="text-sm"
            >
              HOM City
            </div>
            <div
              class="text-xl font-black"
            >
               HOM
            </div>
          </div>
        </a>
        <div
          class="text-sm"
        >
          SOG: 10
        </div>
      </div>
      <div
        class="text-right order-1 sm:order-2"
      >
        <span
          aria-label="HOM City HOM"
          class="w-20 h-20 mx-auto"
          data-mock="team-logo"
          data-team="HOM"
          role="img"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
